### PR TITLE
feat:  automatically share a case with org (FPLA-673)

### PR DESCRIPTION
**Pull request by akomisarek**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/FPLA-673

### Change description ###

Adds functionality to share a case automatically within an organisation. Current implementation is as follows:

1. Check the users from the same localAuthority using existing hardcoded mapping.
2. If the LA is not configured for specified code, check the RD (Reference Data) Professional API for all the user within same organisation. 
3. Adds current's userId to the returned list (if missing) to ensure the code LASOLICITOR role is appropriately configured for user creating a case.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X] No
```

### CODE REVIEW

File: bin/configurer/add-roles.sh [hunk 1]
```diff
@@ -11,15 +11,9 @@ ${dir}/utils/idam-add-role.sh "caseworker-publiclaw"
...
-${dir}/utils/ccd-add-role.sh "caseworker-publiclaw-courtadmin"
-${dir}/utils/idam-add-role.sh "caseworker-publiclaw-cafcass"
-${dir}/utils/ccd-add-role.sh "caseworker-publiclaw-cafcass"
-${dir}/utils/idam-add-role.sh "caseworker-publiclaw-gatekeeper"
-${dir}/utils/ccd-add-role.sh "caseworker-publiclaw-gatekeeper"
-${dir}/utils/idam-add-role.sh "caseworker-publiclaw-systemupdate"
-${dir}/utils/ccd-add-role.sh "caseworker-publiclaw-systemupdate"
-${dir}/utils/idam-add-role.sh "caseworker-publiclaw-judiciary"
-${dir}/utils/ccd-add-role.sh "caseworker-publiclaw-judiciary"
+roles=("-solicitor" "-courtadmin" "-cafcass" "-gatekeeper" "-systemupdate" "-judiciary")
```
Comment (geek-soft) [inline bin/configurer/add-roles.sh]: For me it would look better if role separating dash was removed from here and kept in lines 17-18.

File: README.md [hunk 2]
```diff
@@ -37,16 +37,14 @@ e.g. [direnv](https://direnv.net).
...
-
 ```bash
 $ ./bin/configurer/add-services.sh
 $ ./bin/configurer/add-roles.sh
 $ ./bin/configurer/add-users.sh
-$ ./bin/configurer/utils/generate-local-user-mappings.sh
-
 ```
 
+Users are defined in `bin/configurer/users.json`. Run the final script each time new LA users are added in order to fix access segregation locally.
```
Comment (geek-soft) [inline README.md]: What do you mean by final script? You removed `./bin/configurer/utils/generate-local-user-mappings.sh` from list of scripts above.

Comment (geek-soft) [reply README.md]: Alright, I saw you moved it into `./bin/configurer/add-users.sh`

File: bin/configurer/add-users.sh [hunk 3]
```diff
@@ -4,17 +4,9 @@ set -eu
...
-${dir}/utils/idam-create-caseworker.sh "sam@wiltshire.gov.uk" "caseworker,caseworker-publiclaw,caseworker-publiclaw-solicitor" "(local-authority)"
-${dir}/utils/idam-create-caseworker.sh "hmcts-admin@example.com" "caseworker,caseworker-publiclaw,caseworker-publiclaw-courtadmin" "(hmcts-admin)"
-${dir}/utils/idam-create-caseworker.sh "cafcass@example.com" "caseworker,caseworker-publiclaw,caseworker-publiclaw-cafcass" "(cafcass)"
-${dir}/utils/idam-create-caseworker.sh "judiciary@mailnesia.com" ",caseworker-publiclaw-courtadmin, caseworker,caseworker-publiclaw,caseworker-publiclaw-judiciary,caseworker-publiclaw-gatekeeper" "(judiciary)"
-${dir}/utils/idam-create-caseworker.sh "gatekeeper@mailnesia.com" "caseworker,caseworker-publiclaw,caseworker-publiclaw-gatekeeper,caseworker-publiclaw-courtadmin" "(gatekeeper)"
-${dir}/utils/idam-create-caseworker.sh "fpl-system-update@mailnesia.com" "caseworker,caseworker-publiclaw,caseworker-publiclaw-systemupdate" "(system-update)"
+jq -r '.[] | .email + " " + .roles + " " +  .lastName' ${dir}/users.json | while read args; do
```
Comment (geek-soft) [inline bin/configurer/add-users.sh]: Might be useful to mention in README that `jq` has to be installed (in prerequisites for example). Or maybe use dockerised jq.

File: bin/configurer/users.json [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+  {"email": "raghu@swindon.gov.uk", "roles": "caseworker,caseworker-publiclaw,caseworker-publiclaw-solicitor", "lastName": "(local-authority)"},
+  {"email": "sam@swindon.gov.uk", "roles": "caseworker,caseworker-publiclaw,caseworker-publiclaw-solicitor", "lastName": "(local-authority)"},
+  {"email": "raghu@wiltshire.gov.uk", "roles": "caseworker,caseworker-publiclaw,caseworker-publiclaw-solicitor", "lastName": "(local-authority)"},
+  {"email": "sam@wiltshire.gov.uk", "roles": "caseworker,caseworker-publiclaw,caseworker-publiclaw-solicitor", "lastName": "(local-authority)"},
+  {"email": "hmcts-admin@example.com", "roles": "caseworker,caseworker-publiclaw,caseworker-publiclaw-courtadmin", "lastName": "(hmcts-admin)"},
+  {"email": "cafcass@example.com", "roles": "caseworker,caseworker-publiclaw,caseworker-publiclaw-cafcass", "lastName": "(cafcass)"},
+  {"email": "judiciary@mailnesia.com", "roles": "caseworker-publiclaw-courtadmin,caseworker,caseworker-publiclaw,caseworker-publiclaw-judiciary,caseworker-publiclaw-gatekeeper", "lastName": "(judiciary)"},
```
Comment (geek-soft) [inline bin/configurer/users.json]: List of roles is bit messy here, it starts with more concrete role, then there are more generic roles and then more specific roles again. Every time I need to look into that it gives me headache. Might be an opportunity to clean that up.

File: bin/configurer/utils/generate-local-user-mappings.sh [hunk 5]
```diff
@@ -4,22 +4,50 @@ set -eu
...
+
+function get_users_ids() {
+  query_db "SELECT string_agg(fullobject->>'_id',',') FROM managedObjects WHERE fullobject->>'userName' LIKE '%$1%';"
+}
+
+function create_mock_response() {
+  jq --arg organisation $1 '[.[] | select(.email | contains($organisation))]' $users_file > $mock_file
+  echo $(get_users_email_id_mappings $1) > $users_ids_tmp_file
+
+  for i in `seq 1 $(jq '. | length' $mock_file)`
```
Comment (geek-soft) [inline bin/configurer/utils/generate-local-user-mappings.sh]: Wound't it be easier to start with 0 and not having to do '$index - 1' later?

File: bin/configurer/utils/generate-local-user-mappings.sh [hunk 6]
```diff
@@ -4,22 +4,50 @@ set -eu
...
+  do
+    email=$(jq -r --argjson index $i '.[$index-1].email' $mock_file)
+    user_id=$(jq -r --arg email $email '.[$email]' $users_ids_tmp_file)
+    jq -r --argjson index $i --arg user_id $user_id '.[$index-1].userIdentifier=$user_id | .[$index-1].firstName=.[$index-1].email | .[$index-1].idamStatus="ACTIVE" | .[$index-1].idamStatusCode="0" | .[$index-1].idamMessage="" | .[$index-1].roles|=split(",")' $mock_file > $mock_tmp_file && mv $mock_tmp_file $mock_file
```
Comment (geek-soft) [inline bin/configurer/utils/generate-local-user-mappings.sh]: Not a big fan of this solution. First it is not clear what the end result is cause of all these pipes. There is very strong coupling between input list of users ('users.json') and output mock response. That makes it less readable and unnecessarily more fragile.

File: docker/wiremock/__files/mockResponse.example.json [hunk 7]
```diff
@@ -0,0 +1,16 @@
+[
```
Comment (geek-soft) [inline docker/wiremock/__files/mockResponse.example.json]: Is there any use for that file? I'd easily delete it seeing it is never used. Ideally if script that generates mock response used something like that as a template. That would make it way clearer than modifying JSON objects through 'jq' pipes.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/config/FeignConfiguration.java [hunk 8]
```diff
@@ -12,4 +16,9 @@ public Retryer retryer() {
         return new Retryer.Default(200, 2000, 5);
     }
 
+    @Bean
+    @Primary
+    Decoder feignDecoder(ObjectMapper objectMapper) {
```
Comment (geek-soft) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/config/FeignConfiguration.java]: Out of curiosity why is that needed now?

File: service/src/main/java/uk/gov/hmcts/reform/rd/client/OrganisationApi.java [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+import uk.gov.hmcts.reform.fpl.config.FeignConfiguration;
+import uk.gov.hmcts.reform.rd.model.Status;
+import uk.gov.hmcts.reform.rd.model.User;
+
+import java.util.List;
+
+import static org.springframework.http.HttpHeaders.AUTHORIZATION;
+import static uk.gov.hmcts.reform.ccd.client.CoreCaseDataApi.SERVICE_AUTHORIZATION;
+
+@FeignClient(name = "rd-professional-api", url = "${rd_professional.api.url}", configuration = FeignConfiguration.class)
```
Comment (geek-soft) [inline service/src/main/java/uk/gov/hmcts/reform/rd/client/OrganisationApi.java]: If not tight deadline I would have asked you to move it to separate library so that everyone can reuse it.

File: service/src/main/java/uk/gov/hmcts/reform/rd/client/OrganisationApi.java [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+public interface OrganisationApi {
+    @GetMapping("/refdata/external/v1/organisations/users")
+    List<User> findUsersByOrganisation(
+        @RequestHeader(AUTHORIZATION) String authorisation,
+        @RequestHeader(SERVICE_AUTHORIZATION) String serviceAuthorization,
+        @RequestParam(value = "status") final Status status
```
Comment (geek-soft) [inline service/src/main/java/uk/gov/hmcts/reform/rd/client/OrganisationApi.java]: Make it consistent, either use final keyword or not.

File: service/src/main/java/uk/gov/hmcts/reform/rd/model/User.java [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+import lombok.Builder;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+@Data
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
+public class User {
+    private String userIdentifier;
```
Comment (geek-soft) [inline service/src/main/java/uk/gov/hmcts/reform/rd/model/User.java]: Do you need setter for that class?

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/OrganisationService.java [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+import java.util.List;
+
+@Service
+@Slf4j
+public class OrganisationService {
+    private final LocalAuthorityUserLookupConfiguration localAuthorityUserLookupConfiguration;
+    private final OrganisationApi organisationApi;
+    private final AuthTokenGenerator authTokenGenerator;
+
+    @Autowired
```
Comment (geek-soft) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/OrganisationService.java]: Consider `@RequiredArgsConstructor(onConstructor = @__(@Autowired))` which is quite neat.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/OrganisationService.java [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+    public List<String> findUserIdsInSameOrganisation(String authorisation, String userId, String localAuthorityCode) {
+        List<String> userIds = new ArrayList<>();
+
+        try {
+            addUsersFromSameOrganisationBasedOnAppConfig(localAuthorityCode, userIds);
+        } catch (UnknownLocalAuthorityCodeException ex) {
+            try {
+                addUsersFromSameOrganisationBasedOnReferenceData(authorisation, userIds);
+            } catch (Exception e) {
+                log.warn("Can't find LocalAuthority for code: "
```
Comment (geek-soft) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/OrganisationService.java]: Use logging system support for interpolation  "... {} ..." instead of doing manual string concatenation.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/OrganisationService.java [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+    public OrganisationService(LocalAuthorityUserLookupConfiguration localAuthorityUserLookupConfiguration,
+                               OrganisationApi organisationApi,
+                               AuthTokenGenerator authTokenGenerator) {
+        this.localAuthorityUserLookupConfiguration = localAuthorityUserLookupConfiguration;
+        this.organisationApi = organisationApi;
+        this.authTokenGenerator = authTokenGenerator;
+    }
+
+
+    public List<String> findUserIdsInSameOrganisation(String authorisation, String userId, String localAuthorityCode) {
```
Comment (geek-soft) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/OrganisationService.java]: Not big fan of passing user ID only to add it to list if certain condition is not met. It also add extra responsibility to `findUsers..` method that should be searching for users not having extra responsibilities. I'd rather have handling an empty list scenario outside.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/OrganisationService.java [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+                    + localAuthorityCode
+                    + " in app config and the PRD endpoint call threw an exception.", e);
+            }
+        }
+
+        if (!userIds.contains(userId)) {
+            userIds.add(userId);
+        }
+
+        return userIds;
```
Comment (geek-soft) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/OrganisationService.java]: I am not fan of methods that returns mutable lists. Using list builder and returning immutable list might be clearer.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/OrganisationService.java [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    private void addUsersFromSameOrganisationBasedOnAppConfig(String localAuthorityCode, List<String> userIds) {
```
Comment (geek-soft) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/OrganisationService.java]: Not big fan of methods like that. IMO better and cleaner is to append users in main method with smaller methods that return a list to be appended e.g.

```
private void getUsersFromSameOrganisationBasedOnAppConfig(String localAuthorityCode) { // if you need to rename it for more clarity
  return localAuthorityUserLookupConfiguration.getUserIds(localAuthorityCode)
}

private void getUsersFromSameOrganisationBasedOnReferenceData(String authorisation) {
         return organisationApi
             .findUsersByOrganisation(authorisation, authTokenGenerator.generate(), Status.ACTIVE)
             .stream()
             .map(User::getUserIdentifier)
     }
 }
```

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/OrganisationService.java [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+        this.authTokenGenerator = authTokenGenerator;
+    }
+
+
+    public List<String> findUserIdsInSameOrganisation(String authorisation, String userId, String localAuthorityCode) {
+        List<String> userIds = new ArrayList<>();
+
+        try {
+            addUsersFromSameOrganisationBasedOnAppConfig(localAuthorityCode, userIds);
+        } catch (UnknownLocalAuthorityCodeException ex) {
```
Comment (geek-soft) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/OrganisationService.java]: So you took one or another approach? How will you test it in ITHC env. Will you remove all mapping for FPLA LA and populate PRD database with it?

File: service/src/main/java/uk/gov/hmcts/reform/fpl/config/FeignConfiguration.java (hunk 8 above)
Comment (akomisarek) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/config/FeignConfiguration.java]: To be honest I was not sure why we even needed FeignConfiguration - all the clients we are using (Idam and CaseAccess) is specyfing their own configuration, so I am not sure if it ever worked? Or was it adding the extra retrying functionality to the existing clients? Would need to Debug.

My part is needed as by default it doesn't cope with Json response properly, I would like to explore a bit more as I don't like the NoArgsConstructor still needed, but otherwise it didn't work for me.

File: bin/configurer/add-roles.sh (hunk 1 above)
Comment (pd-dudkiewicz) [reply bin/configurer/add-roles.sh]: good point, changed

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/OrganisationService.java (hunk 12 above)
Comment (akomisarek) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/OrganisationService.java]: Thanks 👍. Implemented

File: bin/configurer/add-users.sh (hunk 3 above)
Comment (pd-dudkiewicz) [reply bin/configurer/add-users.sh]: updated README

File: bin/configurer/users.json (hunk 4 above)
Comment (pd-dudkiewicz) [reply bin/configurer/users.json]: done

File: bin/configurer/utils/generate-local-user-mappings.sh (hunk 5 above)
Comment (pd-dudkiewicz) [reply bin/configurer/utils/generate-local-user-mappings.sh]: agree :+1: changed

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/OrganisationService.java (hunk 17 above)
Comment (akomisarek) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/OrganisationService.java]: I was planning to create new organisation and test the journey e2e for new organisation from scratch (we would still need our mails to be configured but only that).

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/OrganisationService.java (hunk 14 above)
Comment (akomisarek) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/OrganisationService.java]: Ok that's good point. I will refactor.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/OrganisationService.java (hunk 13 above)
Comment (akomisarek) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/OrganisationService.java]: Can I do both exception and placeholders? The SLF4j log API, documentation informs me the arguments I am passing are just objects for placeholders, which makes me nervous.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/OrganisationService.java (hunk 15 above)
Comment (akomisarek) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/OrganisationService.java]: 👍 will refactor.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/OrganisationService.java (hunk 16 above)
Comment (akomisarek) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/OrganisationService.java]: Agree, will reefactor.

File: service/src/main/java/uk/gov/hmcts/reform/rd/model/User.java (hunk 11 above)
Comment (akomisarek) [reply service/src/main/java/uk/gov/hmcts/reform/rd/model/User.java]: Yes, please see my comment in the FeignConfiguration, if time better would investigate a bit more.

File: bin/configurer/utils/generate-local-user-mappings.sh (hunk 6 above)
Comment (pd-dudkiewicz) [reply bin/configurer/utils/generate-local-user-mappings.sh]: Added template file used to fill mock response and therefore reduced number of pipes on this line. But this needs to be strongly coupled with _users.json_ as the idea was to have single place for users definition that's used in all user-related scripts.

File: docker/wiremock/__files/mockResponse.example.json (hunk 7 above)
Comment (pd-dudkiewicz) [reply docker/wiremock/__files/mockResponse.example.json]: Just put it to show how's _mockResponse.json_ supposed to look like. Removed and added _userTemplate.json_ as suggested.

File: service/src/main/java/uk/gov/hmcts/reform/rd/client/OrganisationApi.java (hunk 9 above)
Comment (akomisarek) [reply service/src/main/java/uk/gov/hmcts/reform/rd/client/OrganisationApi.java]: I agree, personally I even think the owning teams could provide the clients, but definitely it should be shared.

File: service/src/main/java/uk/gov/hmcts/reform/rd/client/OrganisationApi.java (hunk 10 above)
Comment (akomisarek) [reply service/src/main/java/uk/gov/hmcts/reform/rd/client/OrganisationApi.java]: 👍

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/OrganisationService.java (hunk 13 above)
Comment (geek-soft) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/OrganisationService.java]: From my experience you can. Exception should be last arg.

File: bin/configurer/utils/generate-local-user-mappings.sh (hunk 6 above)
Comment (geek-soft) [reply bin/configurer/utils/generate-local-user-mappings.sh]: I understand need for having source list of users, what I do not like is that that list, that is used in two or three different places is very strongly coupled with structure of the API that we mock. In that sense we have no freedom of adding anything to elements of the user list because if we do we will smuggle extra stuff to mocked responses. Same way unlikely change in PRD API structure might force us make change in two or three places that are completely unrelated.

In ideal world list of users should have its own structure and shell script should make that structure to structure of PRD API.

Cause of tight deadline I am not asking more for changes. Happy to accept as is ;)

Comment (geek-soft) [reply bin/configurer/utils/generate-local-user-mappings.sh]: Here is 🥇 for you @pd-dudkiewicz for becoming our JQ expert :)

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/OrganisationService.java (hunk 13 above)
Comment (akomisarek) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/OrganisationService.java]: Ok thanks, I have refactored the code so this is no longer relevant, but will remember for the future.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/OrganisationService.java (hunk 15 above)
Comment (akomisarek) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/OrganisationService.java]: Returning ImmutableList now.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/OrganisationService.java (hunk 16 above)
Comment (akomisarek) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/OrganisationService.java]: Refactored.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/LocalAuthorityUserService.java [hunk 18]
```diff
@@ -1,71 +1,75 @@
...
-        List<String> userIds = localAuthorityUserLookupConfiguration.getUserIds(localAuthorityCode);
-
-        if (userIds.isEmpty()) {
-            throw new NoAssociatedUsersException("No users found for the local authority '" + localAuthorityCode + "'");
+    private List<String> findUserIds(String authorisation, String localAuthorityCode) {
+        try {
+            return organisationService
+                .findUserIdsInSameOrganisation(authorisation, localAuthorityCode);
+        } catch (Exception e) {
+            log.warn("Exception while looking for users within the same LA. "
```
Comment (geek-soft) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/LocalAuthorityUserService.java]: I'd say not quite right place to say who gets access but no time to make it better.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/OrganisationService.java [hunk 19]
```diff
@@ -0,0 +1,58 @@
...
+            return ImmutableList
+                .copyOf(getUsersFromSameOrganisationBasedOnAppConfig(localAuthorityCode));
+        } catch (UnknownLocalAuthorityCodeException ex) {
+            try {
+                return
+                    ImmutableList
+                        .copyOf(getUsersFromSameOrganisationBasedOnReferenceData(authorisation));
+            } catch (Exception e) {
+                throw new UserOrganisationLookupException(
+                    format("Can't find users for %s local authority", localAuthorityCode), e
```
Comment (geek-soft) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/service/OrganisationService.java]: Not quite what I asked but template is usually better than concatenation. I am OK with it.
